File: aemet/Aemet-Java-IvanRC/src/main/java/storage/medicion/MedicionStorageServiceJson.java
package storage.medicion;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import config.ConfigApp;
import dto.MapaMedicionesDto;
import mapper.MedicionesMapper;
import models.Medicion;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDate;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class MedicionStorageServiceJson {

    ConfigApp configApp = ConfigApp.getInstance();

    File file = new File(configApp.getAPP_DATA()+File.separator+"mediciones.json");

    Moshi moshi = new Moshi.Builder().build();
    JsonAdapter<MapaMedicionesDto> jsonAdapter = moshi.adapter(MapaMedicionesDto.class);

    public MedicionStorageServiceJson() throws IOException {}

    public void saveAll(Map<LocalDate, List<Medicion>> entities) {
        try(BufferedWriter bw = new BufferedWriter(new FileWriter(file))){
            bw.write((jsonAdapter.indent("   ").toJson(MedicionesMapper.toMedicionesDto(entities))));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public Map<LocalDate, List<Medicion>> loadAll() {
        Map<LocalDate, List<Medicion>> mediciones = Collections.emptyMap();
        if(!file.exists()) return mediciones;
        try{
            String json = Files.readString(Path.of(file.getPath()));
            mediciones = MedicionesMapper.toMediciones(jsonAdapter.fromJson(json));
        }catch (IOException e){
            System.out.println("Error de lectura en ingrediente.json.");
        }
        return mediciones;
    }
}

File: aemet/Aemet-kotlin-IvanRC/src/main/kotlin/storage/medicion/MedicionStorageServiceJson.kt
package storage.medicion

import com.squareup.moshi.Moshi
import com.squareup.moshi.adapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import config.ConfigApp
import dto.MapaMedicionesDto
import mapper.toMediciones
import mapper.toMedicionesDto
import model.Medicion
import mu.KotlinLogging
import java.io.File
import java.time.LocalDate

@ExperimentalStdlibApi
class MedicionStorageServiceJson {

    private val configApp = ConfigApp
    private val file = File(configApp.APP_DATA+File.separator+"mediciones.json")
    private val jsonAdapter = Moshi.Builder().addLast(KotlinJsonAdapterFactory()).build().adapter<MapaMedicionesDto>()

    private val logger = KotlinLogging.logger {  }

    fun saveAll(entities: Map<LocalDate, List<Medicion>>) {
        logger.debug { "Se guardan todas las mediciones sobre el fichero JSON." }
        file.writeText(jsonAdapter.indent("   ").toJson(entities.toMedicionesDto()))
    }

    fun loadAll(): Map<LocalDate, List<Medicion>> {
        logger.debug { "Se cargan todas las mediciones de el fichero JSON." }
        return jsonAdapter.fromJson(file.readText())!!.toMediciones()
    }
}